File: Python/ByTier/F/1774_F2_Magician_and_Pigs_Hard_Version.py
# Problem: CF 1774 F2 - Magician and Pigs (Hard Version)
# https://codeforces.com/contest/1774/problem/F2

"""
Algorithm: Simulation with Lazy Propagation and Binary Search
Time Complexity: O(n * log n)
Space Complexity: O(n)

This solution simulates the given operations on pigs using a combination of
lazy propagation and binary search techniques to efficiently manage
the repeated operations. The key idea is to track how many times each operation
has been applied and use that to determine pig survival counts.

The core logic involves:
1. Tracking cumulative health reductions (s2, s3)
2. Storing operation history to support repeat operations
3. Using binary search via precomputed powers of 2 to efficiently count surviving pigs
"""

import sys

input = sys.stdin.readline

MOD = 998244353
INV2 = MOD // 2 + 1  # Modular inverse of 2
P2 = [pow(2, i, MOD) for i in range(31)]  # Precomputed powers of 2 mod MOD
X = 10**9  # A large value to mark end of operation sequence

s2 = 0  # Cumulative health reduction from type 2 operations
s3 = 0  # Another cumulative health reduction (used to track repetition)
s = []  # List to store cumulative health reductions in a way to simulate repeats
ipt = []  # Stores all the operations in a processed format

# Read number of operations
(n,) = map(int, input().strip().split())

# Process each operation
for _ in range(n):
    i = list(map(int, input().strip().split()))
    op = i[0]
    if op == 1:
        # Create pig with x health points
        ipt.append(i[1])
    elif op == 2:
        # Apply health reduction to all living pigs
        s2 += i[1]
        s3 += i[1]
        ipt.append(-i[1])
    elif op == 3 and s3 < X:
        # Repeat previous operations (simulate)
        s.append(s3)
        s3 *= 2
        ipt.append(0)

ans = 0
s.append(X + 1)  # Sentinel value to mark end
l = 0            # Index tracking of current repetition level
l0 = 0           # Index tracking of base repetition level
while l0 < len(s) and s[l0] == 0:
    l0 += 1
p2l0 = pow(2, l0, MOD)  # Modular power of 2 for base level

for i in ipt:
    if i == 0:
        # Operation was a repeat, update repetition level
        if l < l0:
            p2l0 = p2l0 * INV2 % MOD
        l += 1
    elif i < 0:
        # Health reduction applied, adjust s2
        s2 += i
    elif i > s2:
        # Create a pig and check how many pigs survive based on health
        t = 0
        i -= s2
        r = len(s) - 1
        while r > l and s[r] >= i:
            r -= 1
            if s[r] * 2 < i:
                # Use powers of 2 for fast computation of combinations
                t = (t + P2[r - max(l, l0) + 1]) % MOD
                break
            elif s[r] < i:
                # Add combinations for current level
                t = (t + P2[r - max(l, l0)]) % MOD
                i -= s[r]
        if not (s[r] * 2) < i:
            t += 1
        ans = (ans + t * p2l0) % MOD

print(ans)


# https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/1774/1774_F2_Magician_and_Pigs_Hard_Version.cpp
// Problem: CF 1774 F2 - Magician and Pigs (Hard Version)
// https://codeforces.com/contest/1774/problem/F2

/*
 * Problem: F2. Magician and Pigs (Hard Version)
 * 
 * Purpose:
 *   This code simulates a sequence of operations on pigs' health points and determines the final number of surviving pigs.
 *   The operations include creating pigs, reducing health of all pigs, and repeating all previous operations.
 *   The key idea is to process operations efficiently using a technique that tracks how many times each operation (especially repetition) affects the current state.
 * 
 * Algorithm:
 *   - We simulate the operations but use a compressed representation for repeated operations.
 *   - A state stack is used to represent multiple operations at once:
 *     - Each level of the stack stores:
 *       - sub: cumulative reduction applied to all pigs up to this point
 *       - basis: starting health threshold before this level
 *       - add: list of base health values for pigs created at this level
 *   - Repeating operations cause a new level to be pushed onto the stack.
 *   - The final computation counts surviving pigs using dynamic programming over the operation stack.
 * 
 * Time Complexity: O(n log n)
 *   - Each operation is processed in O(1) time.
 *   - The solve() function processes the accumulated state.
 *   - For repeated operations, the stack depth is bounded by log(n) due to binary representation.
 * 
 * Space Complexity: O(n)
 *   - Stack storage for operation states.
 *   - Additional arrays for tracking counts and intermediate results.
 */
#include <stdio.h>
#include <vector>
#define int long long
using namespace std;
int n, top = 0, ign = 0, w[1000005], counter;
struct {
  int sub, basis;
  vector<int> add;
} state[40];
// Function to compute how many times a health value survives across repeated operations
int count_repr(int val, int i) {
  int ans = 1, j = top;
  while (j >= i) {
    if (val >= state[j].basis) {
      ans += 1 << (j - i);
      val -= state[j].basis;
    }
    j--;
  }
  return ans;
}
// Compute final answer by processing all pig states
void solve() {
  int total = 0, counter = 0;
  for (int i = 0; i <= top; i++)
    for (int val : state[i].add) {
      for (int j = i; j <= top; j++)
        val -= state[j].sub;
      if (val > 0)
        total = (total + w[counter + 1] * count_repr(val - 1, i)) % 998244353;
      counter++;
    }
  printf("%lld", total);
}
signed main() {
  scanf("%lld", &n);
  int current = 0, counter = 0;
  // Initialize the state structure
  for (int i = 0; i < 40; i++)
    state[i] = {0, (int)1e10, {}};
  for (int i = 0; i < n; i++) {
    int op, x;
    scanf("%lld", &op);
    if (op == 1) {
      // Create pig with x health
      scanf("%lld", &x);
      state[top].add.push_back(x + state[top].sub);
      counter++;
    } else if (op == 2) {
      // Reduce all pig health by x
      scanf("%lld", &x);
      state[top].sub += x;
    } else {
      // Repeat all operations until now
      if (ign)
        continue;
      if (top == 0 && state[top].sub == 0) {
        // Special case for repeated operations
        w[counter] = w[counter] == 0 ? 2 : w[counter] * 2 % 998244353;
        continue;
      }
      // Update basis for next level
      state[top].basis =
          state[top].sub + 2 * (top == 0 ? 0 : state[top - 1].basis);
      if (state[top].basis > 2e9) {
        // If basis exceeds limit, ignore further repetitions
        ign = 1;
        continue;
      }
      top++;
    }
  }
  // Precompute powers of 2 for fast counting
  for (int i = counter, v = 1; i >= 0; i--)
    w[i] = w[i] == 0 ? v : v = (w[i] * v) % 998244353;
  solve();
}


// https://github.com/VaHiX/CodeForces/